crates/katana/rpc/rpc/tests/starknet.rs

```diff
@@ -174,3 +179,110 @@ async fn test_send_declare_and_deploy_legacy_contract() {
 
     sequencer.stop().expect("failed to stop sequencer");
 }
+
+#[tokio::test(flavor = "multi_thread")]
+async fn test_messaging_L1_L2() {
+    let root =
+        PathBuf::from(env!("CARGO_MANIFEST_DIR")).join("tests").join("test_data").join("solidity");
+    let paths = ProjectPathsConfig::builder().root(&root).sources(&root).build().unwrap();
+
+    // get the solc project instance using the paths above
+    let project = Project::builder().paths(paths).ephemeral().no_artifacts().build().unwrap();
+
+    // compile the project and get the artifacts
+    let output = project.compile().context("Error compiling project").unwrap();
+    output.assert_success();
+
+    //Get both contracts
+    let contract_starknet_messaging_local =
+        output.find_first("StarknetMessagingLocal").expect("could not find contract").clone();
+
+    let (abi_cstrk, bytecode_cstrk, _) = contract_starknet_messaging_local.into_parts();
+
+    let contract_1 = output.find_first("Contract1").expect("could not find contract").clone();
+
+    let (abi_c1, bytecode_c1, _) = contract_1.into_parts();
+
+    let anvil: TestAnvil = TestAnvil::start().await;
+
+    let account = Arc::new(anvil.account());
+
+    // Factories to deploy the contracts
+    let factory_cstrk =
+        ContractFactory::new(abi_cstrk.clone().unwrap(), bytecode_cstrk.unwrap(), account.clone());
+
+    let factory_c1 =
+        ContractFactory::new(abi_c1.clone().unwrap(), bytecode_c1.unwrap(), account.clone());
+
+    //Deploy to local node (anvil)
+    let contract_strk =
+        factory_cstrk.deploy(()).expect("Failing deploying").send().await.expect("Failing sending");
+
+    assert_eq!(
+        contract_strk.address(),
+        H160::from_str("0x5fbdb2315678afecb367f032d93f642f64180aa3").unwrap()
+    );
+
+    let contract_c1 = factory_c1
+        .deploy(contract_strk.address())
+        .expect("Failing deploying")
+        .send()
+        .await
+        .expect("Failing sending");
+
+    assert_eq!(
+        contract_c1.address(),
+        H160::from_str("0xe7f1725e7734ce288f8367e1bb143e90bb3f0512").unwrap()
+    );
+
+    // -----------------------------------------------------------------------------------------------------------------------------
+
+    let sequencer =
+        TestSequencer::start(SequencerConfig::default(), get_default_test_starknet_config()).await;
+    let account = sequencer.account();
+
+    let path: PathBuf = PathBuf::from("tests/test_data/cairo_l1_msg_contract.json");
+    let (contract, compiled_class_hash) =
+        common::prepare_contract_declaration_params(&path).unwrap();
+
+    let class_hash = contract.class_hash();
+    let res = account.declare(Arc::new(contract), compiled_class_hash).send().await.unwrap();
+
+    // wait for the tx to be mined
+    tokio::time::sleep(Duration::from_millis(WAIT_TX_DELAY_MILLIS)).await;
```

Comment:
Instead of waiting, we could use some transaction waiter here. This will be better than waiting an arbitrary amount of time: https://github.com/dojoengine/dojo/blob/1000fcd6aea4436af7a6e8e93eac865ffd3c2adf/crates/dojo-world/src/utils.rs#L58